src/hotspot/share/gc/g1/g1FullGCOopClosures.inline.hpp

```diff
@@ -73,24 +75,34 @@ template <class T> inline void G1AdjustClosure::adjust_pointer(T* p) {
     return;
   }
 
-  oop forwardee = obj->forwardee();
-  if (forwardee == NULL) {
+  oop forwardee = NULL;
+  if (!ALT_FWD) {
+    forwardee = obj->forwardee();
+  }
+  if ((!ALT_FWD && forwardee == NULL) || (ALT_FWD && !SlidingForwarding::is_forwarded(obj))) {
     // Not forwarded, return current reference.
-    assert(obj->mark_raw() == markOopDesc::prototype_for_object(obj) || // Correct mark
-           obj->mark_raw()->must_be_preserved(obj) || // Will be restored by PreservedMarksSet
-           (UseBiasedLocking && obj->has_bias_pattern_raw()), // Will be restored by BiasedLocking
-           "Must have correct prototype or be preserved, obj: " PTR_FORMAT ", mark: " PTR_FORMAT ", prototype: " PTR_FORMAT,
-           p2i(obj), p2i(obj->mark_raw()), p2i(markOopDesc::prototype_for_object(obj)));
+    if (!ALT_FWD) {
+      assert(obj->mark_raw() == markOopDesc::prototype_for_object(obj) || // Correct mark
+             obj->mark_raw()->must_be_preserved(obj) || // Will be restored by PreservedMarksSet
+             (UseBiasedLocking && obj->has_bias_pattern_raw()), // Will be restored by BiasedLocking
+             "Must have correct prototype or be preserved, obj: " PTR_FORMAT ", mark: " PTR_FORMAT ", prototype: " PTR_FORMAT,
+             p2i(obj), p2i(obj->mark_raw()), p2i(markOopDesc::prototype_for_object(obj)));
+    }
     return;
   }
+  if (ALT_FWD) {
+    forwardee = SlidingForwarding::forwardee<true>(obj);
+  }
```

Comment:
This seems unnecessarily complicated. Can't we put it like:
```
if (ALT_FWD) {
  xxx
} else {
  xxx
}
```

Reply:
updated